src/query/functions/binary/node.go

```diff
@@ -0,0 +1,481 @@
+		}
+	}
+
+	return builder.Build(), nil
+}
+
+// If returnBool is false and op is a comparison function,
+// the function should return the scalar value of lhs if the comparison
+// is true; otherwise, it should return 1 if true, and 0 if false
+func funcReturnBool(fn mathFunc, op string, returnBool bool) mathFunc {
+	if returnBool || !isComparison(op) {
+		return fn
+	}
+	return func(left, right float64) float64 {
+		// NB (arnikola): safe to check that this equals 1 since this is
+		// only performed on comparison functions where viable values
+		// are only 1 or 0
+		take := fn(left, right) == 1
+		if take {
+			return left
+		}
+		// NB(arnikola): whereas prometheus does not include not matching
+		// points in the output, we here replace them with NaNs instead
+		return math.NaN()
+	}
+}
+
+func (n *binaryNode) processBothSeries(
+	lIter, rIter block.StepIter,
+) (block.Block, error) {
+	lMeta, rMeta := lIter.Meta(), rIter.Meta()
+
+	lSeriesMeta := logical.FlattenMetadata(lMeta, lIter.SeriesMeta())
+	rSeriesMeta := logical.FlattenMetadata(rMeta, rIter.SeriesMeta())
+
+	params := n.op.params
+	takeLeft, correspondingRight, lSeriesMeta := intersect(params.VectorMatching, lSeriesMeta, rSeriesMeta)
+
+	lMeta.Tags, lSeriesMeta = logical.DedupeMetadata(lSeriesMeta)
+
+	// Use metas from only taken left series
+	builder, err := n.controller.BlockBuilder(lMeta, lSeriesMeta)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := builder.AddCols(lIter.StepCount()); err != nil {
+		return nil, err
+	}
+
+	returnBoolFn := funcReturnBool(n.op.fn, n.op.OperatorType, params.ReturnBool)
+
+	for index := 0; lIter.Next() && rIter.Next(); index++ {
+		lStep, err := lIter.Current()
+		if err != nil {
+			return nil, err
+		}
+
+		lValues := lStep.Values()
+		rStep, err := rIter.Current()
+		if err != nil {
+			return nil, err
+		}
+
+		rValues := rStep.Values()
+
+		for seriesIdx, lIdx := range takeLeft {
+			rIdx := correspondingRight[seriesIdx]
+			lVal := lValues[lIdx]
+			rVal := rValues[rIdx]
+
+			builder.AppendValue(index, returnBoolFn(lVal, rVal))
+		}
+	}
+
+	return builder.Build(), nil
+}
+
+// computeOrCache figures out if both lhs and rhs are available, if not then it caches the incoming block
+func (n *binaryNode) computeOrCache(ID parser.NodeID, b block.Block) (block.Block, block.Block, error) {
+	var lhs, rhs block.Block
+	n.mu.Lock()
+	defer n.mu.Unlock()
+	op := n.op
+	params := op.params
+	if params.LNode == ID {
+		rBlock, ok := n.cache.Get(params.RNode)
+		if !ok {
+			return lhs, rhs, n.cache.Add(ID, b)
+		}
+
+		rhs = rBlock
+		lhs = b
+	} else if params.RNode == ID {
+		lBlock, ok := n.cache.Get(params.LNode)
+		if !ok {
+			return lhs, rhs, n.cache.Add(ID, b)
+		}
+
+		lhs = lBlock
+		rhs = b
+	}
+
+	return lhs, rhs, nil
+}
+
+func (n *binaryNode) cleanup() {
+	n.mu.Lock()
+	defer n.mu.Unlock()
+	n.cache.Remove(n.op.params.LNode)
+	n.cache.Remove(n.op.params.RNode)
+}
+
+const initIndexSliceLength = 10
```

Comment:
makes sense to move it to the top ? We might want to put some of these things in the config later.